connectors/sources/slack.py

```diff
@@ -0,0 +1,429 @@
+CONVERSATION_HISTORY = "convo_history"
+CONVERSATION_REPLIES = "threads"
+LIST_USERS = "list_users"
+
+ENDPOINTS = {
+    LIST_CONVERSATIONS: "conversations.list",
+    JOIN_CONVERSATIONS: "conversations.join",
+    CONVERSATION_HISTORY: "conversations.history",
+    CONVERSATION_REPLIES: "conversations.replies",
+    LIST_USERS: "users.list",
+}
+
+# TODO list
+# Nice to haves:
+# - expand links
+#   - links:read scope needed?
+# - configure which channels to sync via deny list
+# - write an ftest
+
+
+def retryable_aiohttp_call(retries):  # TODO, make a base client class
+    # TODO: improve utils.retryable to allow custom logic
+    # that can help choose what to retry
+    def wrapper(func):
+        @wraps(func)
+        async def wrapped(*args, **kwargs):
+            retry = 1
+            while retry <= retries:
+                try:
+                    async for item in func(*args, **kwargs):
+                        yield item
+                    break
+                except Exception as e:
+                    if retry >= retries:
+                        raise e
+
+                    retry += 1
+
+        return wrapped
+
+    return wrapper
+
+
+class ThrottledError(Exception):
+    """Internal exception class to indicate that request was throttled by the API"""
+
+    pass
+
+
+class SlackClient:
+    def __init__(self, configuration):
+        self.token = configuration["token"]
+        self._http_session = aiohttp.ClientSession(
+            headers=self._headers(),
+            timeout=aiohttp.ClientTimeout(total=None),
+            raise_for_status=True,
+        )
+        self._sleeps = CancellableSleeps()
+
+    def set_logger(self, logger_):
+        self._logger = logger_
+
+    def ping(self):
+        return
+
+    def close(self):
+        self._http_session.close()
+        return
+
+    async def list_channels(self, only_my_channels, cursor=None):
+        conversation_batch_limit = 200
+        self._logger.debug("Iterating over all channels")
+        if only_my_channels:
+            self._logger.debug("Will only yield channels the bot belongs to")
+        while True:
+            url = f"{BASE_URL}/{ENDPOINTS[LIST_CONVERSATIONS]}?limit={conversation_batch_limit}"
+            if cursor:
+                url = self._add_cursor(url, cursor)
+            response = await self._get_json(url)
+
+            for channel in response["channels"]:
+                if only_my_channels and not channel.get("is_member", False):
+                    continue
+                self._logger.debug(f"Yielding channel '{channel['name']}'")
+                yield channel
+            cursor = self._get_next_cursor(response)
+            if not cursor:
+                break
+
+    async def join_channel(self, channel_id):
+        url = f"{BASE_URL}/{ENDPOINTS[JOIN_CONVERSATIONS]}?channel={channel_id}"
+        return await self._get_json(url)
+
+    async def list_messages(self, channel_id, oldest, latest):
```

Comment:
I see `latest` is always `time.now`, do we want to expose this parameter?

Reply:
It lets us keep the unix timestamp conversion logic in the calling code, and not the client code. 🤷 But I don't feel strongly about it, if you'd prefer not to have the param.